src/main.cpp

```diff
@@ -185,37 +188,67 @@ static int info()
   return 0;
 }
 
-static uint64_t get_btime(void)
+static std::optional<struct timespec> get_boottime()
 {
-  std::ifstream file("/proc/stat");
-  if (!file)
-  {
-    std::cerr << "Fail to open file /proc/stat: " << std::strerror(errno)
-              << std::endl
-              << "Builtin function strftime won't work properly." << std::endl;
-    return 0;
-  }
-  std::string line, field;
-  uint64_t btime = 0;
-  while (std::getline(file, line))
+  std::optional<struct timespec> ret = std::nullopt;
+  long lowest_delta = std::numeric_limits<long>::max();
+
+  // Run the "triple vdso sandwich" 5 times, taking the result from the
+  // iteration with the lowest delta between first and last clock_gettime()
+  // calls.
+  for (int i = 0; i < 5; ++i)
   {
-    std::stringstream ss(line);
-    ss >> field;
-    if (field == "btime")
+    struct timespec before, after, boottime;
+    long delta;
+
+    if (::clock_gettime(CLOCK_REALTIME, &before))
+      continue;
+
+    if (::clock_gettime(CLOCK_BOOTTIME, &boottime))
+      continue;
+
+    if (::clock_gettime(CLOCK_REALTIME, &after))
+      continue;
+
+    // There's no way 3 VDSO calls should take more than 1s. We'll
+    // also ignore the case where we cross a 1s boundary b/c that
+    // can only happen once and we're running this loop 5 times.
+    // This helps keep the math simple.
+    if (before.tv_sec != after.tv_sec)
+      continue;
+
+    delta = after.tv_nsec - before.tv_nsec;
+
+    // Time went backwards
+    if (delta < 0)
+      continue;
+
+    // Lowest delta seen so far, compute boot realtime and store it
+    if (delta < lowest_delta)
     {
-      ss >> btime;
-      if (ss.fail())
-        btime = 0;
-      break;
+      struct timespec boottime_realtime;
+      long nsec_avg = (before.tv_nsec + after.tv_nsec) / 2;
+      if (nsec_avg - boottime.tv_nsec < 0)
+      {
+        boottime_realtime.tv_sec = after.tv_sec - boottime.tv_sec - 1;
+        boottime_realtime.tv_nsec = nsec_avg - boottime.tv_nsec + 1e9;
+      }
+      else
+      {
+        boottime_realtime.tv_sec = after.tv_sec - boottime.tv_sec;
+        boottime_realtime.tv_nsec = nsec_avg - boottime.tv_nsec;
+      }
+
+      lowest_delta = delta;
+      ret = boottime_realtime;
     }
   }
-  if (btime == 0)
-  {
-    std::cerr << "Fail to read btime from /proc/stat. Builtin function "
-                 "strftime won't work properly."
-              << std::endl;
-  }
-  return btime;
+
+  if (ret && lowest_delta >= 1e6)
+    LOG(WARNING) << ">1ms skew detected when calculating boot time. strftime() "
```

Comment:
```suggestion
    LOG(WARNING) << (lowest_delta / 1e3) >> "us skew detected when calculating boot time. strftime() "
                         "builtin may be inaccurate";
```

Reply:
might be useful to know how bad it really is

Reply:
Good idea